test/token_holder/redeem.js

```diff
@@ -0,0 +1,789 @@
+    },
+  );
+
+  const rsv = EthUtils.ecsign(
+    EthUtils.toBuffer(msgHash),
+    EthUtils.toBuffer(ephemeralKey),
+  );
+
+  return { msgHash, rsv };
+}
+
+function getRedeemDataToSign(
+  amount, beneficiary, gasPrice, gasLimit, redeemerNonce) {
+  return web3.eth.abi.encodeFunctionCall(
+    {
+      name: 'redeem',
+      type: 'function',
+      inputs: [
+        {
+          type: 'uint256',
+          name: 'amount',
+        },
+        {
+          type: 'address',
+          name: 'beneficiary',
+        },
+        {
+          type: 'uint256',
+          name: 'gasPrice',
+        },
+        {
+          type: 'uint256',
+          name: 'gasLimit',
+        },
+        {
+          type: 'uint256',
+          name: 'redeemerNonce',
+        },
+      ],
+    },
+    [amount, beneficiary, gasPrice, gasLimit, redeemerNonce],
+  );
+}
+
+async function prepareRedeemPayableRule(
+    accountProvider,
+    amount,
+    beneficiary,
+    facilitator,
+    gasPrice,
+    gasLimit,
+    redeemerNonce,
+    tokenHolder,
+    nonce,
+    ephemeralKey,
+)
+{
+  const mockRule = await MockRule.new();
+
+  const coGatewayRedeemEncodedAbiWithoutHashLock = await getRedeemDataToSign(
+      amount,
+      beneficiary,
+      gasPrice,
+      gasLimit,
+      redeemerNonce
+    );
+
+  const { msgHash, rsv } = getRedeemSignedData(
+    tokenHolder.address,
+    mockRule.address,
+    coGatewayRedeemEncodedAbiWithoutHashLock,
+    nonce,
+    ephemeralKey,
+  );
+
+  return {
+    mockRule,
+    msgHash,
+    rsv,
+  };
+}
+
+async function setupTokenHolder(
+  accountProvider, _ephemeralKeyAddress, _spendingLimit, _deltaExpirationHeight,
+) {
+  token = await UtilityTokenMock.new(1, 1, 'OST', 'Open Simple Token', 1);
+  const tokenRules = await TokenRulesMock.new();
+  const required = 1;
+  const registeredWallet0 = accountProvider.get();
+  const wallets = [registeredWallet0];
+
+  const tokenHolder = await TokenHolder.new(
+    token.address,
+    tokenRules.address,
+    wallets,
+    required,
+  );
+
+  const blockNumber = await web3.eth.getBlockNumber();
+
+  await tokenHolder.submitAuthorizeSession(
+    _ephemeralKeyAddress,
+    _spendingLimit,
+    blockNumber + _deltaExpirationHeight,
+    { from: registeredWallet0 },
+  );
+
+  return {
+    tokenHolder,
+    registeredWallet0,
+  };
+}
+
+contract('TokenHolder::redeem', async (accounts) => {
+
+  describe('Negative Tests', async () => {
+    const accountProvider = new AccountProvider(accounts);
+    const spendingLimit = 50,
+      deltaExpirationHeight = 100,
+      amount = 10,
+      beneficiary = accountProvider.get(),
+      facilitator = accountProvider.get(),
+      gasPrice = 10,
+      gasLimit = 10,
+      redeemerNonce = 1,
+      hashLock = web3.utils.soliditySha3('hl'),
+      payableValue = 100,
+      nonce = 1;
+
+    it('Reverts if ExTx is signed with non-authorized key.', async () => {
+
+      const {
+        tokenHolder,
+      } = await setupTokenHolder(
```

Comment:
I think setupTokenHolder is already an utility function. It's looking longer because of larger parameters it expects . Also return parameters for setupTokenHolder is different for different its. Also I think readability wise some duplication is fine in test cases.